script/tool/lib/src/drive_examples_command.dart

```diff
@@ -43,213 +48,252 @@ class DriveExamplesCommand extends PluginCommand {
-                targetPaths
-                    .add(p.relative(integrationTest.path, from: example.path));
-              }
-            }
-
-            if (targetPaths.isEmpty) {
-              print('''
-Unable to infer a target application for $driverTestName to drive.
-Tried searching for the following:
-1. test/$deviceTestName
-2. test_driver/$deviceTestName
-3. test_driver/*_test.dart
-''');
-              failingTests.add(p.relative(test.path, from: example.path));
-              continue;
-            }
-          }
-
-          final List<String> driveArgs = <String>['drive'];
-
-          final String enableExperiment = getStringArg(kEnableExperiment);
-          if (enableExperiment.isNotEmpty) {
-            driveArgs.add('--enable-experiment=$enableExperiment');
-          }
-
-          if (isLinux && isLinuxPlugin(plugin)) {
-            driveArgs.addAll(<String>[
-              '-d',
-              'linux',
-            ]);
-          }
-          if (isMacos && isMacOsPlugin(plugin)) {
-            driveArgs.addAll(<String>[
-              '-d',
-              'macos',
-            ]);
-          }
-          if (isWeb && isWebPlugin(plugin)) {
-            driveArgs.addAll(<String>[
-              '-d',
-              'web-server',
-              '--web-port=7357',
-              '--browser-name=chrome',
-            ]);
-          }
-          if (isWindows && isWindowsPlugin(plugin)) {
-            driveArgs.addAll(<String>[
-              '-d',
-              'windows',
-            ]);
-          }
-
-          for (final String targetPath in targetPaths) {
-            testsRan = true;
-            final int exitCode = await processRunner.runAndStream(
-                flutterCommand,
-                <String>[
-                  ...driveArgs,
-                  '--driver',
-                  p.join('test_driver', driverTestName),
-                  '--target',
-                  targetPath,
-                ],
-                workingDir: example,
-                exitOnError: true);
-            if (exitCode != 0) {
-              failingTests.add(p.join(packageName, deviceTestPath));
-            }
-          }
+
+        testsRan = true;
+        final List<File> failingTargets = await _driveTests(
+            example, driver, testTargets,
+            deviceFlags: deviceFlags);
+        for (final File failingTarget in failingTargets) {
+          errors.add(p.relative(failingTarget.path, from: package.path));
         }
       }
-      if (!testsRan) {
-        pluginsWithoutTests.add(pluginName);
-        print(
-            'No driver tests run for $pluginName ($examplesFound examples found)');
-      }
     }
-    print('\n\n');
+    if (!testsRan) {
+      printError('No driver tests were run ($examplesFound example(s) found).');
+      errors.add('No tests ran (use --exclude if this is intentional).');
+    }
+    return errors;
+  }
+
+  Future<List<String>> _getDevicesForPlatform(String platform) async {
```

Comment:
nit: Splitting this into parsing the results of `devices` and calling `devices` would make it more testable.  (Although it looks like you can use MockProcess to test it fine.)

Reply:
For now relying on the indirect testing is fine; if we end up needing this in more places and pull this out as a utility method we can split it apart and do direct unit testing.